src/main/java/gmailgdogra/SwipeProcessor.java

```diff
@@ -1,84 +1,91 @@
-                List.of("Thames LSI0904 - Turnstile North OUT",
-                        "Thames LSI0902 - Turnstile South OUT"));
-        allSwipeOutDevices.put(Location.VISITORS_RECEPTION,
-                List.of("Thames LSI0302 - Empl W Turnstile OUT",
-                        "Thames LSI0304 - Empl E Turnstile OUT"));
-        allSwipeOutDevices.put(Location.EP_WEIGHBRIDGE,
-                List.of("Thames LSI0702 - Weighbridge OUT"));
-        allSwipeOutDevices.put(Location.TL_PLAISTOW,
-                List.of("PLA0103 - Turnstile West OUT",
-                        "PLA0105 - Turnstile East OUT"));
+        allSwipeOutDevices.add("Thames LSI0904 - Turnstile North OUT");
+        allSwipeOutDevices.add("Thames LSI0902 - Turnstile South OUT");
+        allSwipeOutDevices.add("Thames LSI0302 - Empl W Turnstile OUT");
+        allSwipeOutDevices.add("Thames LSI0304 - Empl E Turnstile OUT");
+        allSwipeOutDevices.add("Thames LSI0702 - Weighbridge OUT");
+        allSwipeOutDevices.add("PLA0103 - Turnstile West OUT");
+        allSwipeOutDevices.add("PLA0105 - Turnstile East OUT");
 
     }
 
     public static List<SwipeRecord> getOutputDataFrom(List<SwipeRecord> allSwipes) {
 
         List<SwipeRecord> outputData = new ArrayList<>();
-        Set<Officer> officers = UserInput.getOfficers(allSwipes);
-        for (Officer officer : officers) {
-            if (officer.getLocation() != null) {
-                outputData.add(getFirstSwipeIn(officer, allSwipes)); // swipe in row
-                outputData.add(getLastSwipeOut(officer, allSwipes)); // swipe out row
-            }
+        List<SwipeRecord> onlyInOutSwipes = filterInOutSwipes(allSwipes);
+        Set<Officer> allOfficers = ExtractOfficers.from(onlyInOutSwipes);
+        List<Shift> shifts = UserInput.getShiftDetails(allOfficers);
+
+        for (Shift shift : shifts) {
+            outputData.add(getSwipeIn(shift, onlyInOutSwipes));
+            outputData.add(getSwipeOut(shift, onlyInOutSwipes));
         }
         return outputData;
     }
 
-    public static SwipeRecord getFirstSwipeIn(Officer officer, List<SwipeRecord> allSwipes) {
-        Location loc = officer.getLocation();
+    private static List<SwipeRecord> filterInOutSwipes(List<SwipeRecord> allSwipes) {
         return allSwipes.stream()
-                .filter(swipe -> officer.getFullName().equals(swipe.getFullName()))
-                .filter(swipe -> allSwipeInDevices.get(loc).contains(swipe.getDeviceName()))
+                .filter(swipe -> allSwipeInDevices.contains(swipe.getDeviceName()) ||
+                        allSwipeOutDevices.contains(swipe.getDeviceName()))
+                .collect(Collectors.toList());
+    }
+
+    private static SwipeRecord getSwipeIn(Shift shift, List<SwipeRecord> onlyInOutSwipes) {
+        return onlyInOutSwipes.stream()
+                .filter(swipe -> swipe.getFullName().equals(shift.getOfficer().getFullName()))
```

Comment:
Be warned that in the case of any error is spelling you will get unpredicted behaviour. I don't know if you sanitize personal names somewhere and somehow (e.g. all capitals, extra space removed, etc.), but keep this in mind.

Reply:
the user does not input the name in any process. so no concerns on this part.